### PR TITLE
rec: Do not put results of DS query for auth or forward domains in negcache.

### LINKED ISSUE #10189: rec: Consider ways to handle DS queries for names from export-etc-hosts or forward-zones
- Program: Recursor
 - Issue type: Bug report

### Short description
`DS` queries can cause issues with names served from `export-etc-hosts` or `forward-zones`. Various combinations of DNSSEC and other options could potentially make this worse (untested).

This demonstrates it for `export-etc-hosts` but the same conditions can occur when using `forward-zones`

### Environment
 - Operating system: Debian 
 - Software version: 4.5.0~alpha3+master.101.g5d9b7eaf6-1pdns.buster
 - Software source: PowerDNS repository

### Steps to reproduce

Setup a minimal environment using `export-etc-hosts`

```
root@slurpee2:/etc/powerdns# systemctl restart pdns-recursor
root@slurpee2:/etc/powerdns# cat recursor.conf 
config-dir=/etc/powerdns
hint-file=/usr/share/dns/root.hints
include-dir=/etc/powerdns/recursor.d
local-address=127.0.0.1
lua-config-file=/etc/powerdns/recursor.lua
quiet=yes
setgid=pdns
setuid=pdns

dnssec=off
root-nx-trust=no
qname-minimization=no
nothing-below-nxdomain=no

export-etc-hosts=yes
root@slurpee2:/etc/powerdns# grep powerdns /etc/hosts
127.0.53.53 powerdns.corp
root@slurpee2:/etc/powerdns# dig ds powerdns.corp @127.0.0.1

; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> ds powerdns.corp @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 8133
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;powerdns.corp.			IN	DS

;; AUTHORITY SECTION:
.			3600	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2021031800 1800 900 604800 86400

;; Query time: 85 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu Mar 18 09:34:36 MDT 2021
;; MSG SIZE  rcvd: 117

root@slurpee2:/etc/powerdns# dig powerdns.corp @127.0.0.1

; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> powerdns.corp @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 28933
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;powerdns.corp.			IN	A

;; AUTHORITY SECTION:
.			3597	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2021031800 1800 900 604800 86400

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu Mar 18 09:34:39 MDT 2021
;; MSG SIZE  rcvd: 117

root@slurpee2:/etc/powerdns# 
```

### Expected behaviour
Expect the A record to always be served, and not trashed by the preceding DS query.

### Actual behaviour
See above.

### Other information
We solve this issue by using a `preresolve` hook, however this seems to have hit a few folks so documented here incase there are some thoughts on it.

### PR DESCRIPTION
**Pull request by omoerbeek**
Fixes #10189, and does not break existing unit tests.

Needs unit test.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by rgacogne** [comment]
Is it still enough if the aggressive NSEC cache is enabled?

**Comment by omoerbeek** [comment]
> Is it still enough if the aggressive NSEC cache is enabled?

A quick test with validation (and thus aggressive NSEC cache) enabled shows it works for the case in the issue. The test for looking into the aggressive NSEC cache (bottom of `doCacheCkeck()`is:
```
if (g_aggressiveNSECCache && !wasForwardedOrAuthZone) ...
```
So that hints it is not used for this case.

**Comment by rgacogne** [comment]
Great :)

File: pdns/syncres.cc [hunk 1]
```diff
@@ -3418,9 +3424,16 @@ bool SyncRes::processRecords(const std::string& prefix, const DNSName& qname, co
         continue;
       }
     }
+    const bool negCacheIndiction = rec.d_place == DNSResourceRecord::AUTHORITY && rec.d_type == QType::SOA &&
+      lwr.d_rcode == RCode::NXDomain && qname.isPartOf(rec.d_name) && rec.d_name.isPartOf(auth);
+
+    bool putInNegCache = true;
+    if (negCacheIndiction && isForwardOrAuth(qname)) {
```
Comment (rgacogne) [inline pdns/syncres.cc:3431]: Did you consider skipping the insertion only when the response comes from a DS query? For auth zones that probably doesn't matter but I'm wondering if we really want to disable this even when the response comes from the forwarded server?
Untested:
```suggestion
    if (negCacheIndiction && isForwardOrAuth(qname) && qtype == QType::DS) {
```

Comment (omoerbeek) [reply pdns/syncres.cc:3431]: Ugh, I had that in my head but neglected  to type that extra condition

**Review by rgacogne — APPROVED** [review]

**Pull request merged**